spec source for provisional extensions (#695)
KhronosGroup/OpenCL-Docs: ext/cl_khr_external_semaphore.asciidoc

depending on the handle's type. The import method used is referred to as the
handle type's transference. When using handle types with reference transference,
importing a payload to a semaphore adds the semaphore to the set of all
semaphores sharing that payload. This set includes the semaphore from which the
payload was exported. Semaphore signaling and waiting operations performed on
any semaphore in the set must behave as if the set were a single semaphore.
Importing a payload using handle types with copy transference creates a
duplicate copy of the payload at the time of import, but makes no further
reference to it. Semaphore signaling and waiting operations performed on the
target of copy imports must not affect any other semaphore or payload.

Export operations have the same transference as the specified handle type's
import operations. Additionally, exporting a semaphore payload to a handle with
copy transference has the same side effects on the source semaphore's payload as
executing a semaphore wait operation. If the semaphore was using a temporarily
imported payload, the semaphore's prior permanent payload will be restored.

> Please, refer to handle specific specs for more details on transference and

Reviewer: ```suggestion
Please refer to handle specific specifications for more details on transference and
```

The `cl_khr_external_semaphore_dx_fence` extension extends {cl_external_semaphore_handle_type_khr_TYPE} to support the following new types of handles, and adds as a property that may be specified when creating a semaphore from an external handle:

--
    * {CL_SEMAPHORE_HANDLE_D3D12_FENCE_KHR} specifies an NT handle returned by ID3D12Device::CreateSharedHandle referring to a Direct3D 12 fence, or ID3D11Device5::CreateFence referring to a Direct3D 11 fence. It owns a reference to the underlying synchronization primitive associated with the Direct3D fence.
--

When waiting on semaphores using {clEnqueueWaitSemaphoresKHR} or signaling semaphores using {clEnqueueSignalSemaphoresKHR}, the semaphore payload must be provided for semaphores created from {CL_SEMAPHORE_HANDLE_D3D12_FENCE_KHR}.

* If _sema_objects_ list has a mix of cl_semaphore obtained from *CL_SEMAPHORE_HANDLE_D3D12_FENCE_KHR* and other handle types,
then the _sema_payload_list_ should point to a list of _num_sema_objects_ payload values for each cl_semaphore in _sema_objects_. 
However, the payload values corresponding to semaphores with type CL_SEMAPHORE_TYPE_BINARY_KHR can be set to 0 or will be ignored.

> *clEnqueueWaitSemaphoresKHR* and *clEnqueueSignalSemaphoresKHR* may return *CL_INVALID_VALUE* if _sema_objects_ list has one or more cl_semaphore obtained from *CL_SEMAPHORE_HANDLE_D3D12_FENCE_KHR* and _sema_payload_list is NULL.

Reviewer: ```suggestion
*clEnqueueWaitSemaphoresKHR* and *clEnqueueSignalSemaphoresKHR* may return *CL_INVALID_VALUE* if _sema_objects_ list has one or more cl_semaphore obtained from *CL_SEMAPHORE_HANDLE_D3D12_FENCE_KHR* and _sema_payload_list_ is NULL.
```

--

Transference and permanence properties for handle types added by `cl_khr_external_semaphore_win32`:

.Transference and Permanence Properties for `cl_khr_external_semaphore_win32` handles
[width="100%",cols="60%,<20%,<20%",options="header"]
|====
| Handle Type | Transference | Permanence
| *CL_SEMAPHORE_HANDLE_OPAQUE_WIN32_KHR*
  |  Reference
      | Temporary, Permanent
| *CL_SEMAPHORE_HANDLE_OPAQUE_WIN32_KMT_KHR*
  |  Reference
      | Temporary, Permanent
|====

For these extensions, importing a semaphore payload from Windows handles does not transfer ownership of the handle to the OpenCL implementation. For handle types defined as NT handles, the application must release ownership using the CloseHandle system call when the handle is no longer needed.

[[cl_khr_external_semaphore-Sample-Code]]
=== Sample Code

> . Example for importing semaphore created by other API in OpenCL in single-device context.

Reviewer: ```suggestion
. Example for importing a semaphore created by another API in OpenCL in single-device context.
```

+
--
[source]
----
// Get cl_devices of the platform.
clGetDeviceIDs(..., &devices, &deviceCount);

// Create cl_context with just first device
clCreateContext(..., 1, devices, ...);
    
// Obtain fd/win32 or similar handle for external semaphore to be imported

> // from other API.

Reviewer: ```suggestion
// from the other API.
```

int fd = getFdForExternalSemaphore();

// Create clSema of type cl_semaphore_khr usable on only available device0
// assuming the semaphore was imported from the same device.

cl_semaphore_properties_khr sema_props[] = 
        {(cl_semaphore_properties_khr)CL_SEMAPHORE_TYPE_KHR,
         (cl_semaphore_properties_khr)CL_SEMAPHORE_TYPE_BINARY_KHR,
         (cl_semaphore_properties_khr)CL_SEMAPHORE_HANDLE_OPAQUE_FD_KHR,
         (cl_semaphore_properties_khr)fd,
          0};

int errcode_ret = 0;
cl_semaphore_khr clSema = clCreateSemaphoreWithPropertiesKHR(context,
                                                             sema_props,
                                                             &errcode_ret);

----
--

> . Example for importing semaphore created by other API in OpenCL in multi-device context for single device usage.

Reviewer: ```suggestion
. Example for importing a semaphore created by another API in OpenCL in a multi-device context for single device usage.
```

+
--
[source]
----
// Get cl_devices of the platform.
clGetDeviceIDs(..., &devices, &deviceCount);

// Create cl_context with first two devices
clCreateContext(..., 2, devices, ...);
    
// Obtain fd/win32 or similar handle for external semaphore to be imported

> // from other API.

Reviewer: ```suggestion
// from the other API.
```

int fd = getFdForExternalSemaphore();

> // Create clSema of type cl_semaphore_khr usable only on device1

Reviewer: ```suggestion
// Create clSema of type cl_semaphore_khr usable only on devices[1]
```

// assuming the semaphore was imported from the same device.

cl_semaphore_properties_khr sema_props[] = 
        {(cl_semaphore_properties_khr)CL_SEMAPHORE_TYPE_KHR, 
         (cl_semaphore_properties_khr)CL_SEMAPHORE_TYPE_BINARY_KHR,
         (cl_semaphore_properties_khr)CL_SEMAPHORE_HANDLE_OPAQUE_FD_KHR, 
         (cl_semaphore_properties_khr)fd,
         (cl_semaphore_properties_khr)CL_DEVICE_HANDLE_LIST_KHR, 
         (cl_semaphore_properties_khr)devices[1], CL_DEVICE_HANDLE_LIST_END_KHR,
          0};

int errcode_ret = 0;
cl_semaphore_khr clSema = clCreateSemaphoreWithPropertiesKHR(context,
                                                             sema_props,
                                                             &errcode_ret);

----
--

> . Example for synchronization using semaphore created by other API and imported in OpenCL

Reviewer: ```suggestion
. Example for synchronization using a semaphore created by another API and imported in OpenCL
```

+
--
[source]
----

> // Create clSema usinge one of the above examples of external semaphore creation.

Reviewer: ```suggestion
// Create clSema using one of the above examples of external semaphore creation.
```

                                                             &errcode_ret);

// Start the main loop

while (true) {
    // (not shown) Signal the semaphore from the other API

    // Wait for the semaphore in OpenCL
    clEnqueueWaitSemaphoresKHR(/*command_queue*/           command_queue,
                               /*num_sema_objects*/        1,
                               /*sema_objects*/            &clSema,
                               /*num_events_in_wait_list*/ 0,
                               /*event_wait_list*/         NULL,
                               /*event*/                   NULL);

    // Launch kernel
    clEnqueueNDRangeKernel(command_queue, ...);

    // Signal the semaphore in OpenCL
    clEnqueueSignalSemaphoresKHR(/*command_queue*/           command_queue,
                                 /*num_sema_objects*/        1,
                                 /*sema_objects*/            &clSema,
                                 /*num_events_in_wait_list*/ 0,
                                 /*event_wait_list*/         NULL,
                                 /*event*/                   NULL);

>     // (not shown) Launch work in other API that waits on  'clSema'

Reviewer: ```suggestion
    // (not shown) Launch work in the other API that waits on 'clSema'
```

                                   sizeof(int),
                                   &fd,
                                   NULL);
}

// Start the main rendering loop

while (true) {
    // (not shown) Signal the semaphore from the other API

    // Wait for the semaphore in OpenCL
    clEnqueueWaitSemaphoresKHR(/*command_queue*/           command_queue,
                               /*num_sema_objects*/        1,
                               /*sema_objects*/            &clSema,
                               /*num_events_in_wait_list*/ 0,
                               /*event_wait_list*/         NULL,
                               /*event*/                   NULL);

    // Launch kernel
    clEnqueueNDRangeKernel(command_queue, ...);

    // Signal the semaphore in OpenCL
    clEnqueueSignalSemaphoresKHR(/*command_queue*/           command_queue,
                                 /*num_sema_objects*/        1,
                                 /*sema_objects*/            &clSema,
                                 /*num_events_in_wait_list*/ 0,
                                 /*event_wait_list*/         NULL,
                                 /*event*/                   NULL);

>     // (not shown) Launch work in other API that waits on  'clSema'

Reviewer: ```suggestion
    // (not shown) Launch work in the other API that waits on 'clSema'
```
